### ISSUE TITLE
FR: encrypt export

### ISSUE THREAD

**Issue by romeokienzler** [helloworld1/FreeOTPPlus#128]
Encrypt export using e.g. OpenPGP or OpenSSL library. Otherwise the file can be read and stolen as residing on the file system.

Workaround: I encrypt the file after export using Openkeychain and then upload it to a cloud provider for backup. But while the file resides on the FS unencrypted it can be read by other apps.

**Comment by Minoru** [comment]
Isn't this a duplicate of #122, @romeokienzler?

**Comment by OmlineEditor** [comment]
encrypting files for export is important and necessary.

**Comment by helloworld1** [comment]
Agreed. I plan to use a symmetric encryption with user password to encrypt all the app data and exports. 
The main issue is the backward compatibility. I need to make sure to also support unencrypted data if user choose to, and also during app upgrade.

**Comment by OmlineEditor** [comment]
do encryption even if the phone does not have a screen lock at all and you only need to turn it on to access the phone. please take this into account when designing

**Comment by helloworld1** [comment]
I hope the encryption will be completely independently from system lock /account meaning you can have encryption and password without a screen lock.
The biometrics can be used to login faster.

**Comment by romeokienzler** [comment]
@helloworld1 sorry for the duplicate and the late reply - did you have a chance to look into the issue? just created a backup and encrypted with "openkeychain" - little concerned that the json is lying around unencrypted for 30-60 seconds :)

**Comment by helloworld1** [comment]
@romeokienzler Encryption is considered. There are some major change in the codebase to make the storage more flexible. The new encryptionType field in db will be used to indicate if the data is encrypted or not
https://github.com/helloworld1/FreeOTPPlus/blob/master/token-data/src/main/java/org/fedorahosted/freeotp/data/OtpToken.kt#L19

**Comment by PhilKes** [comment]
Any news on this issue?
If not I would try to implement a proposition for symmetric encryption with a user password (AES Encryption).
Any thoughts/suggestions on that?

**Comment by helloworld1** [comment]
It's not being implemented now. Feel free to contribute. The room database has a field indicating the row is encrypted or not. Currently all are plain text though.

The integration of fingerprint authentication is the challenging part. Probably we can do in a separate PR.

**Comment by 8ware** [comment]
Since getting encryption/decryption right is a [difficult](https://github.com/helloworld1/FreeOTPPlus/pull/214#issuecomment-1641033344) [task](https://github.com/helloworld1/FreeOTPPlus/pull/214#event-12968668852) I suggest to provide a share intent instead to pipe the export into OpenKeychain for encryption. That way, FreeOTP+ doesn't need to handle encryption while the resulting file export can be protected when hitting the shared storage. Likewise, the app could register as [receiver](https://developer.android.com/training/sharing/send#using-the-right-mimetype) for importing OpenKeychain-decrypted backups.

One drawback is that the export could be shared accidentally with someone in your messaging app as it will probably shown up in Android's Sharesheet by default. To mitigate this the intent could be explicitly directed to OpenKeychain using its [` org.sufficientlysecure.keychain.action.ENCRYPT_TEXT `](https://github.com/open-keychain/open-keychain/wiki/Intents#openkeychain-intent-actions) intent. However, the other way around (i.e. from OpenKeychain to FreeOTP+) will be hard(er) to enforce.

What do you think?
